docs/workspaces.md

```diff
@@ -89,7 +89,8 @@ To configure one or more `Workspaces` in a `Task`, add a `workspaces` list with
   
 Note the following:
   
-- A `Task` definition can include as many `Workspaces` as it needs. 
+- A `Task` definition can include as many `Workspaces` as it needs. It is recommended that `Tasks` use
+  **at most** one _writable_ `Workspace`.
```

Comment:
Thanks for reviewing!

>Is this true for any kind of workspace, regardless of the type of volume backing them?

I write _recommend_ since it is not strictly true, but for a Task to be usable in most clusters. 

We have workspace volume sources: `Secret`, `ConfigMap`, `PersistentVolumeClaim` and `emptyDir`. I consider `Secret` and `ConfigMap` as _readonly_. You can have multiple writeable `emptyDir` in your `Task` - but not useful in a `PipelineRun`.

You can have multiple writeable `PersistentVolumeClaim` if it supports [access mode](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) `ReadWriteMany` - but there are few storage solutions available with that access mode, typically e.g. NFS-servers. If using this, this feature can be disabled.

PVCs with access mode `ReadOnlyMany` can be mounted on multiple Nodes at the time, so it works. But is **read-only**. I should probably update my code to not add Affinity Assistant in those cases.

Using PVC with access mode `ReadWriteOnce` (the most common access mode) can only be mounted on one Node at a time. In addition, they most commonly lives in a single datacenter/Availability Zone. So if a pipeline starts e.g. with two parallel tasks - they may be scheduled to two different Availability Zones - this goes fine - until a Task try to mount both volumes - located in different AZ, and the pipeline is deadlocked - but this only happens when scheduled to different zones. This happened to me in the [parallel example](https://github.com/tektoncd/pipeline/pull/2546) that I added to the examples and caused _flaky_ tests for the Tekton Pipelines project - and had to **remove**. I also documented all these technicalities in [Specifying workspace order in a pipeline](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md#specifying-workspace-order-in-a-pipeline) but that section is full of technicalities and even a **warning**. This was my main motivation to create this feature - in this PR that section of the documentation is less technical and the corner-cases is solved with this PR.

But as you say, you _can_ use multiple writable volumes, but you need to be careful and know what you are doing.

The improved performance with this PR was a side-effect, my main motivation was to make it easy to use commonly availably PVCs in parallel without deadlocks or Tasks that are timed out (as in the current warning).

I think it is good that we _recommend_ to design Tasks with only one writeable workspace - this makes it fully functional in all cases when using this feature. There is probably other good solutions for cases that first was designed with multiple writable workspaces? e.g. using buckets or other storage that is not limited to only one AZ and is not _mounted_ on the Node?

This is a complex field were I run into many problems the last weeks, we should improve the documentation as you say. For me it is also important to make Tekton easy to use without corner-cases in a non-technical way :)

Reply:
> Thanks for reviewing!
> 
> > Is this true for any kind of workspace, regardless of the type of volume backing them?
> 
> I write _recommend_ since it is not strictly needed, but for a Task to be usable in most clusters.
> 
> We have workspace volume sources: `Secret`, `ConfigMap`, `PersistentVolumeClaim` and `emptyDir`. I consider `Secret` and `ConfigMap` as _readonly_. You can have multiple writeable `emptyDir` in your `Task` - but not useful in a `PipelineRun`.
> 
> You can have multiple writeable `PersistentVolumeClaim` if it supports [access mode](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes) `ReadWriteMany` - but there are few storage solutions available with that access mode, typically e.g. NFS-servers. If using this, this feature can be disabled.
> 
> PVCs with access mode `ReadOnlyMany` can be mounted on multiple Nodes at the time, so it works. But is **read-only**. I should probably update my code to not add Affinity Assistant in those cases.

+1 - could also be a follow-up PR if you want
you'll need to rebase in any case

> 
> Using PVC with access mode `ReadWriteOnce` (the most common access mode) can only be mounted on one Node at a time. In addition, they most commonly lives in a single datacenter/Availability Zone. So if a pipeline starts e.g. with two parallel tasks - they may be scheduled to two different Availability Zones - this goes fine - until a Task try to mount both volumes - located in different AZ, and the pipeline is deadlocked - but this only happens when scheduled to different zones. This happened to me in the [parallel example](https://github.com/tektoncd/pipeline/pull/2546) that I added to the examples and caused _flaky_ tests for the Tekton Pipelines project - and had to **remove**. I also documented all these technicalities in [Specifying workspace order in a pipeline](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md#specifying-workspace-order-in-a-pipeline) but that section is full of technicalities and even a **warning**. This was my main motivation to create this feature - in this PR that section of the documentation is less technical and the corner-cases is solved with this PR.
> 
> But as you say, you _can_ use multiple writable volumes, but you need to be careful and know what you are doing.
> 
> The improved performance with this PR was a side-effect, my main motivation was to make it easy to use commonly availably PVCs in parallel without deadlocks or Tasks that are timed out (as in the current warning).
> 
> I think it is good that we _recommend_ to design Tasks with only one writeable workspace - this makes it fully functional in all cases when using this feature. There is probably other good solutions for cases that first was designed with multiple writable workspaces? e.g. using buckets or other storage that is not limited to only one AZ and is not _mounted_ on the Node?

I agree on the recommendation, I was just wondering if we should document some of the reasoning behind it, or what issues one might run into when using more than one.

> 
> This is a complex field were I run into many problems the last weeks, we should improve the documentation as you say. For me it is also important to make Tekton easy to use without corner-cases in a non-technical way :)

Reply:
Yes, I'll take more extensive documentation about PVCs and what to think about in a separate PR.